Assets/Scripts/Assignment/CostBasedAssignment.cs

```diff
@@ -0,0 +1,65 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+// The cost-based assignment assigns hierarchical objects to each other based on the pairwise costs
+// between them.
+public class CostBasedAssignment : AssignmentBase {
+  // Delegate for determining the cost of an assignment between two hierarchical objects.
+  public delegate float CostDelegate(IHierarchical first, IHierarchical second);
+
+  // Delegate for performing the assignment.
+  public delegate Plugin.StatusCode AssignDelegate(int numFirst, int numSecond, float[] costs,
+                                                   int[] assignedFirsts, int[] assignedSeconds,
+                                                   out int numAssignments);
+
+  // Maximum cost to prevent overflow.
+  private const float _maxCost = 1e12f;
+
+  private readonly CostDelegate _costFunction;
+  private readonly AssignDelegate _assignFunction;
+
+  public CostBasedAssignment(CostDelegate costFunction, AssignDelegate assignFunction) {
+    _costFunction = costFunction;
+    _assignFunction = assignFunction;
+  }
+
+  // Run the assignment algorithm and assign the hierarchical objects.
+  public override List<AssignmentItem> Assign(IReadOnlyList<IHierarchical> first,
+                                              IReadOnlyList<IHierarchical> second) {
+    int numFirst = first.Count;
+    int numSecond = second.Count;
+    if (numFirst == 0 || numSecond == 0) {
+      return new List<AssignmentItem>();
+    }
+
+    // Find all pairwise assignment costs.
+    var assignmentCosts = new float[numFirst * numSecond];
+    for (int firstIndex = 0; firstIndex < numFirst; ++firstIndex) {
+      for (int secondIndex = 0; secondIndex < numSecond; ++secondIndex) {
+        float cost = _costFunction(first[firstIndex], second[secondIndex]);
+        assignmentCosts[firstIndex * numSecond + secondIndex] =
+            Mathf.Clamp(cost, -_maxCost, _maxCost);
+      }
+    }
+
+    // Solve the assignment problem.
+    var assignedFirstIndices = new int[numFirst];
+    var assignedSecondIndices = new int[numFirst];
+    Plugin.StatusCode status =
+        _assignFunction(numFirst, numSecond, assignmentCosts, assignedFirstIndices,
+                        assignedSecondIndices, out int numAssignments);
+    if (status != Plugin.StatusCode.StatusOk) {
+      Debug.Log($"Failed to run the assignment with status code {status}.");
+      return new List<AssignmentItem>();
+    }
```

Comment:
We should probably LogError here not just Log since this would be fatal?


```suggestion
    if (status != Plugin.StatusCode.StatusOk) {
      Debug.LogError($"Assignment failed with status {status}. " +
                        $"Input: {numFirst} first, {numSecond} second. " +
                        $"Cost range: [{assignmentCosts.Min()}, {assignmentCosts.Max()}]");
      return new List<AssignmentItem>();
    }
```